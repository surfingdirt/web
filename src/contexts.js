import Cookies from 'js-cookie';
import React from 'react';
import Login from '~/Login';
import routes from '~/routes';

const { HOME } = routes;
const LoginCookie = Login.COOKIE_NAME;
const signupDestinationKey = 'signUpDestination';

class AppContextValueObject {
  constructor(values) {
    this.values = {
      SSR: false,
      accountKey: '',
      availableLanguages: [],
      baseUrl: '',
      dir: 'ltr',
      facebookAppId: '',
      googleClientId: '',
      twitterUsername: '',
      graphql: '',
      language: '',
      login: {
        data: {
          accessToken: null,
          me: {
            id: null,
            username: '',
            email: '',
            firstname: '',
            lastname: '',
            gender: '',
            birthYear: '',
            country: '',
            phone: '',
            imageUrl: '',
            status: '',
            favoritesTeam: [],
          },
        },
      },
      nowValue: null,
      screenWidth: undefined,
      sportPlayer: { iframe: '', src: '' },
      translations: {},
    };

    Object.keys(values).forEach((key) => {
      if (this.values.hasOwnProperty(key)) {
        this.values[key] = values[key];
      } else {
        throw new Error(`Trying to assign unknown property: ${key}`);
      }
    });

    this.values.login.onSuccess = ({ accessToken, expires }) => {
      Cookies.set(LoginCookie, accessToken, { expires });

      const destination = this.values.login.getOrigin() || HOME;
      this.values.login.clearOrigin();
      window.location = destination;
    };

    this.values.login.onFailure = () => {
      this.values.login.data = {
        id: null,
        accessToken: null,
        expires: null,
        tokenType: null,
      };
    };

    this.values.login.logout = () => {
      Cookies.remove(LoginCookie);
      this.resetUser();
    };

    this.values.login.saveOrigin = () => {
      if (!global.window) {
        // Don't worry about that on the server.
        return;
      }
      sessionStorage.setItem(signupDestinationKey, window.location.pathname);
    };

    this.values.login.clearOrigin = () => {
      sessionStorage.removeItem(signupDestinationKey);
    };

    this.values.login.getOrigin = () => {
      if (!global.window) {
        // Don't worry about that on the server.
        return null;
      }
      return sessionStorage.getItem(signupDestinationKey);
    };
  }

  setAccessToken(accessToken) {
    this.values.login.data.accessToken = accessToken;
  }

  setUser({ id, username, email, bookmarks, likes, watches }) {
    this.values.login.data.me = { id, username, email, bookmarks, likes, watches };
  }

  resetUser() {
    this.setUser({
      id: null,
      username: '',
      email: '',
    });
  }

  getValues() {
    return this.values;
  }
}

const AppContext = React.createContext(AppContextValueObject);

export default {
  AppContext,
  AppContextValueObject,
};
