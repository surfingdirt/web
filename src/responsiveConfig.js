// This list should be the only place to add/modify breakpoint information for
// responsive design. Storybook, page-level media queries use this configuration.
const MOBILE_LOW = 'MOBILE_LOW';
const MOBILE_HI = 'MOBILE_HI';
const DESKTOP_LOW = 'DESKTOP_LOW';
const DESKTOP_HI = 'DESKTOP_HI';

const entries = [
  {
    id: MOBILE_LOW,
    displayName: 'Low-end mobile',
    width: 320,
    height: 480,
  },
  {
    id: MOBILE_HI,
    displayName: 'High-end mobile',
    width: 360,
    height: 640,
  },
  {
    id: DESKTOP_LOW,
    displayName: 'Low-end desktop',
    width: 1024,
    height: 768,
  },
  {
    id: DESKTOP_HI,
    displayName: 'High-end desktop',
    width: 1920,
    height: 1080,
  },
];

export const getClassNameForWidth = (width) => `screenwidth-${width}-only`;

export const standardScreenWidths = {};
export const storybookViewports = {};
export const screenWidthClassNames = {};

entries.forEach((entry) => {
  const { id, width, height, displayName } = entry;

  standardScreenWidths[id] = width;
  storybookViewports[id] = {
    name: displayName,
    styles: {
      width: `${width}px`,
      height: `${height}px`,
    },
    width,
    height,
  };
  screenWidthClassNames[id] = getClassNameForWidth(width);
});

export const getActiveScreenWidth = (screenWidth) => {
  const widths = Object.values(standardScreenWidths).sort((a, b) => a - b);

  let activeWidth = widths[0];
  widths.forEach((width) => {
    if (screenWidth >= width) {
      activeWidth = width;
    }
  });

  return activeWidth;
};

/**
 * TODO: figure out a way to make these rules dynamic (SASS), and constructed at build time.
 */
export const mediaQueryRules = {
  [standardScreenWidths[MOBILE_LOW]]: `
`,

  [standardScreenWidths[MOBILE_HI]]: `
`,

  [standardScreenWidths[DESKTOP_LOW]]: `
  .limitedWidthContent {
    width: 960px;
  }
`,

  [standardScreenWidths[DESKTOP_HI]]: `
  .limitedWidthContent {
    width: 1464px;
  }
  
  .fullWidthContent {
    width: 1900px;
  }
`,
};
