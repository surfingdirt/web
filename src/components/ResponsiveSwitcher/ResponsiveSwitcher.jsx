import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import AppContext from '~/contexts';
import { getActiveScreenWidth, getClassNameForWidth } from '~/responsiveConfig';

/*
 * Example
 * <ResponsiveSwitcher widthToComponentMap={...}>
 *	<ResponsiveVersionRed />
 *	<ResponsiveVersionGreen />
 *	<ResponsiveVersionBlue />
 * </ResponsiveSwitcher>
 */

export default class ResponsiveSwitcher extends React.PureComponent {
  // noinspection JSUnusedGlobalSymbols
  static contextType = AppContext;

  static propTypes = {
    children: PropTypes.node.isRequired,
    widthToComponentMap: PropTypes.objectOf(PropTypes.func).isRequired,
  };

  constructor(props) {
    super(props);

    const screenWidth = typeof window === 'undefined' ? 0 : window.innerWidth;

    this.state = { activeScreenWidth: getActiveScreenWidth(screenWidth) };

    this.resizeListener = this.resizeListener.bind(this);
  }

  componentDidMount() {
    window.addEventListener('resize', this.resizeListener);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.resizeListener);
  }

  getPropsForChild(width, index) {
    // Filter out certain props before passing them on
    const childProps = {};
    const blacklist = ['children', 'widthToComponentMap'];
    const { props } = this;

    Object.keys(props).forEach((key) => {
      if (blacklist.includes(key)) {
        return;
      }
      childProps[key] = props[key];
    });

    childProps.key = `key-${index}`;

    return childProps;
  }

  getKey(i) {
    return `key-${i}`;
  }

  findComponentInList(list, activeComponent) {
    let component = null;
    React.Children.forEach(list, (currentChild) => {
      if (currentChild.type === activeComponent) {
        component = currentChild;
      }
    });

    if (!component) {
      throw new Error(
        `Could not find component of type ${activeComponent} among the ResponsiveSwitcher direct descendants`,
      );
    }

    return component;
  }

  resizeListener() {
    this.setState({ activeScreenWidth: getActiveScreenWidth(window.innerWidth) });
  }

  renderChildWithContainer(width, component, i) {
    return (
      <div key={this.getKey(i)} className={classnames(getClassNameForWidth(width))}>
        {React.cloneElement(component, this.getPropsForChild(width, i))}
      </div>
    );
  }

  renderChildren(children, widthToComponentMap, clientSide, activeScreenWidth) {
    const $children = [];

    const activeComponent = widthToComponentMap[activeScreenWidth] || null;

    Object.entries(widthToComponentMap).forEach(([widthAsString, currentComponent], i) => {
      const width = parseInt(widthAsString, 10);
      let $child;

      if (clientSide) {
        if (width !== activeScreenWidth) {
          // On the client, render unnecessary components as empty, hidden divs.
          $child = <div hidden key={this.getKey(i)} />;
        } else {
          const component = this.findComponentInList(children, activeComponent);
          $child = this.renderChildWithContainer(width, component, i);
        }
      } else {
        const component = this.findComponentInList(children, currentComponent);
        $child = this.renderChildWithContainer(width, component, i);
      }

      $children.push($child);
    });
    return $children;
  }

  render() {
    const { SSR } = this.context;

    const { activeScreenWidth } = this.state;
    const { children, widthToComponentMap } = this.props;

    return this.renderChildren(children, widthToComponentMap, !SSR, activeScreenWidth);
  }
}
