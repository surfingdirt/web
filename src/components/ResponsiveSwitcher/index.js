import ResponsiveSwitcher from './ResponsiveSwitcher';

export default ResponsiveSwitcher;
