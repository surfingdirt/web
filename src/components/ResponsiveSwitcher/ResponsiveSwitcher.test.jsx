import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import contexts from '~/contexts';
import { standardScreenWidths } from '~/responsiveConfig';

import ResponsiveSwitcher from './ResponsiveSwitcher';

const { MOBILE_LOW, DESKTOP_LOW } = standardScreenWidths;

// noinspection JSUnusedLocalSymbols
const { AppContext } = contexts;

const VersionA = ({ className }) => <div className={classnames(className, 'version-a')} />;
VersionA.propTypes = { className: PropTypes.string };
VersionA.defaultProps = { className: '' };

const VersionB = ({ className }) => <div className={classnames(className, 'version-b')} />;
VersionB.propTypes = { className: PropTypes.string };
VersionB.defaultProps = { className: '' };

const widthToComponentMap = {
  [MOBILE_LOW]: VersionA,
  [DESKTOP_LOW]: VersionB,
};

const getRenderedMarkup = (context, map) =>
  mount(
    <AppContext.Provider value={context}>
      <ResponsiveSwitcher widthToComponentMap={map}>
        <VersionA />
        <VersionB />
      </ResponsiveSwitcher>
    </AppContext.Provider>,
  );

describe('ResponsiveSwitcher', () => {
  it('should render the server-side version of the markup', () => {
    const wrapper = getRenderedMarkup({ SSR: true }, widthToComponentMap);
    expect(wrapper).toMatchSnapshot();

    // VersionA is rendered within an element with class screenwidth-320-only
    const version320 = wrapper.find('.screenwidth-320-only');
    expect(version320).toHaveLength(1);
    const versionA = version320.find('.version-a');
    expect(versionA).toHaveLength(1);

    // VersionB is rendered within an element with class screenwidth-1024-only
    const version1024 = wrapper.find('.screenwidth-1024-only');
    expect(version1024).toHaveLength(1);
    const versionB = version1024.find('.version-b');
    expect(versionB).toHaveLength(1);
  });

  it('should render client-side markup for a low-end mobile device', () => {
    global.innerWidth = 320;

    const wrapper = getRenderedMarkup({ SSR: false }, widthToComponentMap);
    expect(wrapper).toMatchSnapshot();

    const divs = wrapper.children();

    const div320 = divs.at(0);
    expect(div320.hasClass('screenwidth-320-only')).toBeTruthy();
    const versionA = div320.find('.version-a');
    expect(versionA).toHaveLength(1);

    // 1024 element is hidden
    expect(divs.at(1).getDOMNode().hidden).toBeTruthy();
  });

  it('should render client-side markup for a low-end mobile device if the resolution is below minimum', () => {
    global.innerWidth = 100;

    const wrapper = getRenderedMarkup({ SSR: false }, widthToComponentMap);
    expect(wrapper).toMatchSnapshot();

    const divs = wrapper.children();

    const div320 = divs.at(0);
    expect(div320.hasClass('screenwidth-320-only')).toBeTruthy();
    const versionA = div320.find('.version-a');
    expect(versionA).toHaveLength(1);

    // 1024 element is hidden
    expect(divs.at(1).getDOMNode().hidden).toBeTruthy();
  });

  it('should render client-side markup for a high-end desktop device', () => {
    global.innerWidth = 1024;

    const wrapper = getRenderedMarkup({ SSR: false }, widthToComponentMap);
    expect(wrapper).toMatchSnapshot();

    const divs = wrapper.children();
    // 320 element is hidden
    expect(divs.at(0).getDOMNode().hidden).toBeTruthy();

    const div1024 = divs.at(1);
    expect(div1024.hasClass('screenwidth-1024-only')).toBeTruthy();
    const versionB = div1024.find('.version-b');
    expect(versionB).toHaveLength(1);
  });
});
