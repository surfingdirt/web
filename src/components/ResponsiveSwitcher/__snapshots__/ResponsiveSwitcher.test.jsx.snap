// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsiveSwitcher should render client-side markup for a high-end desktop device 1`] = `
Array [
  <div
    hidden={true}
    key="key-0"
  />,
  <div
    className="screenwidth-1024-only"
    key="key-1"
  >
    <div
      className="version-b"
    />
  </div>,
]
`;

exports[`ResponsiveSwitcher should render client-side markup for a low-end mobile device 1`] = `
Array [
  <div
    className="screenwidth-320-only"
    key="key-0"
  >
    <div
      className="version-a"
    />
  </div>,
  <div
    hidden={true}
    key="key-1"
  />,
]
`;

exports[`ResponsiveSwitcher should render client-side markup for a low-end mobile device if the resolution is below minimum 1`] = `
Array [
  <div
    className="screenwidth-320-only"
    key="key-0"
  >
    <div
      className="version-a"
    />
  </div>,
  <div
    hidden={true}
    key="key-1"
  />,
]
`;

exports[`ResponsiveSwitcher should render the server-side version of the markup 1`] = `
Array [
  <div
    className="screenwidth-320-only"
    key="key-0"
  >
    <div
      className="version-a"
    />
  </div>,
  <div
    className="screenwidth-1024-only"
    key="key-1"
  >
    <div
      className="version-b"
    />
  </div>,
]
`;
